### PR TITLE
🌱 Added go directive test

### LINKED ISSUE #10258: Add presubmit check to prevent accidental go directive bump
In Cluster API we only bump the go directive in go.mod files on the main branch.

While we bump the Go minor version that we build release branches with we never bump the go directive there.

This issue is about adding a presubmit check to ensure we don't accidentally bump it.

This can be implemented by adding a new `verify-go-directive` Makefile target (prior art: https://github.com/kubernetes/utils/pull/306/files) and then adding it to the `verify` target we already run in CI.

For more context, see: https://github.com/kubernetes/kubernetes/issues/123744

### PR DESCRIPTION
**Pull request by AvineshTripathi**
fixes #10258

### CODE REVIEW

**Comment by fabriziopandini** [comment]
thanks for picking up this task!
lgtm pending fixing the shell-check error

File: Makefile [hunk 1]
```diff
@@ -686,6 +686,10 @@ ALL_VERIFY_CHECKS = licenses boilerplate shellcheck tiltfile modules gen convers
 .PHONY: verify
 verify: $(addprefix verify-,$(ALL_VERIFY_CHECKS)) lint-dockerfiles ## Run all verify-* targets
 
+.PHONY: verify-go-directive
+verify-go-directive:
+	TRACE=$(TRACE) ./hack/verify-go-directive.sh -g 1.21
```
Comment (killianmuldoon) [inline Makefile:691]: Could we try to parameterize this? there's a `GO_VERSION` variable in this file

File: .github/workflows/pr-verify.yaml [hunk 2]
```diff
@@ -12,8 +12,20 @@ jobs:
...
+      - name: Verifier action
+        id: verifier
+        uses: kubernetes-sigs/kubebuilder-release-tools@012269a88fa4c034a0acf1ba84c26b195c0dbab4 # tag=v0.4.3
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Install Go
+        uses: actions/setup-go@v3
+        with:
+          go-version: 1.21
```
Comment (killianmuldoon) [inline .github/workflows/pr-verify.yaml:24]: We should be able to paramaterize this - see https://github.com/kubernetes-sigs/cluster-api/blob/3c86142d92a7dc9bf84f3991731de1c73bd9755c/.github/workflows/pr-golangci-lint.yaml#L25-L26

Comment (sbueringer) [reply .github/workflows/pr-verify.yaml:24]: Let's please not add it to this GitHub action.

We already have a `verify` Makefile target. Let's run it there, i.e. let's add it to the list in ~l.684 in Makefile (ALL_VERIFY_CHECKS)

File: Makefile (hunk 1 above)
Comment (sbueringer) [reply Makefile:691]: Not really because the GO_VERSION in the makefile and the one in the go directive are not always in sync (e.g. on release branches)

File: .github/workflows/pr-verify.yaml [hunk 3]
```diff
@@ -12,8 +12,20 @@ jobs:
     runs-on: ubuntu-latest
     name: verify PR contents
     steps:
-    - name: Verifier action
-      id: verifier
-      uses: kubernetes-sigs/kubebuilder-release-tools@012269a88fa4c034a0acf1ba84c26b195c0dbab4 # tag=v0.4.3
-      with:
-        github_token: ${{ secrets.GITHUB_TOKEN }}
+      - name: Verifier action
```
Comment (sbueringer) [inline .github/workflows/pr-verify.yaml:15]: Please revert the changes to this file (see https://github.com/kubernetes-sigs/cluster-api/pull/10261/files#r1522991772)

File: hack/verify-go-directive.sh [hunk 4]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env bash
+
+# Copyright 2014 The Kubernetes Authors.
```
Comment (sbueringer) [inline hack/verify-go-directive.sh:3]:
```suggestion
# Copyright 2024 The Kubernetes Authors.
```

**Review by sbueringer — COMMENTED** [review]
Otherwise lgtm

**Comment by AvineshTripathi** [comment]
@sbueringer am not sure if my changes are breaking stuff could you confirm

File: Makefile [hunk 5]
```diff
@@ -681,11 +681,15 @@ APIDIFF_OLD_COMMIT ?= $(shell git rev-parse origin/main)
...
 
-ALL_VERIFY_CHECKS = licenses boilerplate shellcheck tiltfile modules gen conversions doctoc capi-book-summary diagrams import-restrictions
+ALL_VERIFY_CHECKS = licenses boilerplate shellcheck tiltfile modules gen conversions doctoc capi-book-summary diagrams import-restrictions go-directive
 
 .PHONY: verify
 verify: $(addprefix verify-,$(ALL_VERIFY_CHECKS)) lint-dockerfiles ## Run all verify-* targets
 
+.PHONY: verify-go-directive
+verify-go-directive:
+	TRACE=$(TRACE) ./hack/verify-go-directive.sh -g 1.21
```
Comment (fabriziopandini) [inline Makefile:691]: Let's add a GO_DIRECTIVE_VERSION near to GO_VERSION

**Review by fabriziopandini — COMMENTED** [review]
One nit, otherwise lgtm

**Comment by fabriziopandini** [comment]
/test pull-cluster-api-test-main

This seems a flake (there is also a discussion thread in [slack](https://kubernetes.slack.com/archives/C8TSNPY4T/p1710728340192279))

File: Makefile (hunk 5 above)
Comment (fabriziopandini) [reply Makefile:691]: Instead of having the hardcoded const 1.21 in L691 (deep down the Makefile), let's introduce a new variable named `GO_DIRECTIVE_VERSION=1.21` at the beginning of the Makefile, just below the definition of GO_VERSION and GO_CONTAINER_IMAGE.
If we do in this way, all the go version will be defined in a single place easier to find and maintain

**Comment by fabriziopandini** [comment]
@AvineshTripathi thanks for working on this!
/lgtm
cc @sbueringer @killianmuldoon for a final pass (otherwise I will merge in a few days)

**Review by killianmuldoon — COMMENTED** [review]
Thanks @AvineshTripathi!

/lgtm
/approve
/hold

in case someone else want to have another look.

**Comment by sbueringer** [comment]
Would be great if we can manual cherry-pick this PR into 1.6 & 1.5 (in both cases we should use 1.20 as GO_DIRECTIVE_VERSION)

@AvineshTripathi Do you want to take this over?

**Comment by sbueringer** [comment]
Basically you create the same PR against the release-1.6 & release-1.5 branch (but using 1.20 as GO_DIRECTIVE_VERSION). You can either do it entirely manually or something like this:

```bash
git checkout release-1.6
git checkout -b release-1.6-go-directive
git cherry-pick <commit of this PR>
# Change to 1.20
git commit --amend
...
```

Just saying "Manually" because usually we have a cherry-pick bot. But this won't be enough as we want to use 1.20
